src/MacVim/MMCoreTextView.m

```diff
@@ -127,26 +126,98 @@ - (void)drawInvertedRectAtRow:(int)row column:(int)col numRows:(int)nrows
     return [@"m" sizeWithAttributes:a].width;
 }
 
-@implementation MMCoreTextView
+typedef struct {
+    unsigned color;
+    int shape;
+    int fraction;
+} GridCellInsertionPoint;
+
+typedef struct {
+    // Note: All objects should be weak references.
+    // Fields are grouped by draw order.
+    BOOL inverted;
+
+    // 1. Background
+    unsigned bg;
+
+    // 2. Sign
+    NSImage* sign;
+
+    // 3. Insertion point
+    GridCellInsertionPoint insertionPoint;
+
+    // 4. Text
+    unsigned fg;
+    unsigned sp;
+    int textFlags;
+    CTFontRef font;
+    CGGlyph glyph;
+} GridCell;
+
+typedef struct {
+    GridCell *cells;
+    int rows;
+    int cols;
+} Grid;
+
+static GridCell* grid_cell(Grid *grid, int row, int col) {
+    return grid->cells + row * grid->cols + col;
+}
+
+// Returns a static cell if row or col is out of bounds. Draw commands can point
+// out of bounds if -setMaxRows:columns: is called while Vim is still drawing at
+// a different size, which has been observed when exiting non-native fullscreen
+// with `:set nofu`. If that gets fixed, then delete this workaround.
+static GridCell* grid_cell_safe(Grid *grid, int row, int col) {
+    if (row >= grid->rows || col >= grid->cols) {
+        static GridCell scratch_cell;
+        return &scratch_cell;
+    }
+    return grid_cell(grid, row, col);
+}
+
+static void grid_resize(Grid *grid, int rows, int cols) {
+    const int oldSize = grid->rows * grid->cols;
+    const int newSize = rows * cols;
+    if (rows == grid->rows && cols == grid->cols)
+        return;
+    if (grid->cells == NULL) {
+        grid->cells = calloc(rows * cols, sizeof(GridCell));
+    } else {
+        if (cols < grid->cols) {
```

Comment:
I believe there is a bug here. If you let's say resize from 1 row x 5 cols to 5 rows x 1 col, `newSize` will be the same as `oldSize`, and the old data won't get properly cleared out. Basically, `grid->cells[1..4]` will have junk data (they should be zero), while `grid->cells[0]` is correct (it has the value of first row first column).

I think this bug doesn't manifest itself because Vim is just going to redraw everything but we should probably fix this in case other code tries to use it later.

Reply:
Fixed and simplified.